app/lib/storage/encryption.ts

```diff
@@ -0,0 +1,159 @@
+import * as crypto from "crypto"
+
+/**
+ * This module contains all the functions related to encryption middleware for
+ * storage.
+ */
+
+/**
+ * This interface represents the arguments of the Options class constructor.
+ */
+export interface IProtoOptions {
+  algorithm?: string
+  digest?: string
+  iterations?: number
+  iv?: Buffer
+  ivBytes?: number
+  keyBytes?: number
+  password: string
+  salt?: Buffer
+  saltBytes?: number
+}
+
+/**
+ * The Options class bundles together all cryptographic settings and methods for
+ * a vault.
+ */
+export class Options {
+  /** The name of a cryptographic algorithm */
+  public algorithm: string
+  /** The name of a hash function */
+  public digest: string
+  /** Number of PBKDF2 iterations */
+  public iterations: number
+  /** PBKDF2 initialization vector */
+  public iv: Buffer
+  /** PBFDK2 initialization vector byte length */
+  public ivBytes: number
+  /** AES key byte length */
+  public keyBytes: number
+  /** User defined password */
+  public password: string
+  /** PBFDK2 salt */
+  public salt: Buffer
+  /** PBFDK2 salt byte length */
+  public saltBytes: number
+
+  constructor(opts: IProtoOptions) {
+    this.algorithm = opts.algorithm || "aes-256-cbc"
+    this.digest = opts.digest || "sha256"
+    this.iterations = opts.iterations || 100000
+    this.ivBytes = opts.ivBytes || 16
+    this.keyBytes = opts.keyBytes || 32
+    this.password = opts.password
+    this.saltBytes = opts.saltBytes || 32
+
+    this.iv = opts.iv || crypto.randomBytes(this.ivBytes)
+    this.salt = opts.salt || crypto.randomBytes(this.saltBytes)
+  }
+
+  /**
+   * Derive the PBKDF2 key for current options and a certain salt.
+   * @param {Buffer} salt
+   * @returns {Buffer}
+   */
+  public getKey(salt: Buffer = this.salt): Buffer {
+    return crypto.pbkdf2Sync(
+      this.password,
+      salt,
+      this.iterations,
+      this.keyBytes,
+      this.digest
+    )
+  }
+}
+
+/**
+ * Crypto buffers serializer function.
+ * @param {Array<Buffer>} buffers
+ * @returns {Buffer}
+ */
+const serialize = (buffers: Array<Buffer>): Buffer => {
+  const parts: Array<Buffer> = []
+
+  buffers.forEach( (part: Buffer) => {
+    const len = Buffer.alloc(4)
+    len.writeUInt32BE(part.length, 0)
+    parts.push(len)
+    parts.push(part)
+  })
+
+  return Buffer.concat(parts)
+}
+
+/**
+ * Crypto buffers deserializer function.
+ * @param {Buffer} data
+ * @returns {Array<Buffer>}
+ */
+const deserialize = (data: Buffer): Array<Buffer> => {
+  const parts: Array<Buffer> = []
+  const l = data.length
```

Comment:
for clarity, the name of the variable 'l' could be changed to something like 'len'